### PR TITLE
feat(blog): 4.12 Article

### PR DESCRIPTION
**Pull request by Princesseuh**

### CODE REVIEW

File: src/content/blog/astro-4120.mdx [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+TODO: Matthew Please Write This Part
+
+## Added `first` and `last` URLs to pagination data
+
+The [paginate() helper](https://docs.astro.build/en/reference/api-reference/#paginate) now returns the `first` and `last` properties in the pagination data. These properties contain the first and last page urls, respectively. Thanks to [@tsawada](https://github.com/tsawada) for contributing this feature!
+
+## Added support for the Shiki's `defaultColor` option
+
+The `defaultColor` option allows you to override the values of a theme's inline style, adding only CSS variables to give you more flexibility in applying multiple color themes. This option can be configured in your Shiki config to apply throughout your site, or passed to Astro's built-in `<Code>` component to style an individual code block. Thank you to [@madcampos](https://github.com/madcampos) for contributing this!
```
Comment (sarah11918) [inline src/content/blog/astro-4120.mdx:48]:
``````suggestion
The `defaultColor` option allows you to override the values of a theme's inline style, adding only CSS variables to give you more flexibility in applying multiple color themes.

This option can be configured in your Shiki config to apply throughout your site, or passed to Astro's built-in `<Code>` component to style an individual code block. Thank you to [@madcampos](https://github.com/madcampos) for contributing this!

```js title="astro.config.mjs"
import { defineConfig } from 'astro/config';
export default defineConfig({
  markdown: {
    shikiConfig: {
      themes: {
        light: 'github-light',
        dark: 'github-dark',
      },
      defaultColor: false,
    },
  },
});
```

```astro
---
import { Code } from 'astro:components';
---
<Code code={`const useMyColors = true`} lang="js" defaultColor={false} />
```
``````
Totally not necessary, just throwing these here in case you liked having the examples visually to break up text! People like to skim for code!

Comment (Princesseuh) [reply src/content/blog/astro-4120.mdx:48]: I *do* like it, however I've deliberately kept the sections that are not "Server Islands" small-ish so for marketing-purposes, so not sure!

Comment (sarah11918) [reply src/content/blog/astro-4120.mdx:48]: Entirely your call! That does make sense! You might have a better sense after the Server Islands part is added whether it feels right.

File: src/content/blog/astro-4120.mdx [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+  output: 'hybrid',
+  adapter: netlify(),
+
+  experimental: {
+    serverIslands: true,
+  },
+})
+```
+
+Server islands, as the name implies, run on the server, so you'll need to use either `'server'` or `'hybrid'` output. Once you've enabled the feature you can use them on any component using the `server:defer` directive:
```
Comment (florian-lefebvre) [inline src/content/blog/astro-4120.mdx:80]: With this sentence I understood: you can use `server:defer` and `client:load` at the same time

```astro
<Foo server:defer client:load />
```
Is that what you wanted to say?

File: src/content/blog/astro-4120.mdx [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+As you can see from this table, the numbers are extremely close. Server Islands has an advantage in painting slightly faster due to being more lightweight both in the infrastructure setting in front of the request, and the amount of code needed in the client throughout the process. Partial prerendering has an advantage in that it can finish slightly faster, due to starting the request for its deferred content on the server before the HTML reaches the user's browser.
+
+Overall we are happy with the trade-offs we've made for Server Islands. Our implementation happens mostly at build-time, where component content is swapped out for a small script, one that can almost fit into a Tweet. This means that Server Islands aren't dependant on any particular server infrastructure and will run on host.
+
+### Getting involved
+
+Server Islands has an [active RFC](https://github.com/withastro/roadmap/pull/963). We're still finalizing the details and features; we'll likely be adding encrypted props for example. Try out the release in 4.12 and let us know what you think.
```
Comment (florian-lefebvre) [inline src/content/blog/astro-4120.mdx:110]:
```suggestion
Server Islands have an [active RFC](https://github.com/withastro/roadmap/pull/963). We're still finalizing the details and features; we'll likely be adding encrypted props for example. Try out the release in 4.12 and let us know what you think.
```

File: src/content/blog/astro-4120.mdx [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+```astro
+<Avatar server:defer>
+  <GenericUserImage slot="fallback" />
+</Avatar>
+```
+
+When Astro builds your site it will omit the component and inject a script in its place, in addition to the content you've marked with `slot="fallback"`. When the page loads in the browser these components will be requested, along with the props you passed to them, to a special endpoint that renders them and returns the HTML.
+
+Server Islands are normal Astro components and can use any of the features you expect from Astro like middleware, slots, `client:` islands.
```
Comment (florian-lefebvre) [inline src/content/blog/astro-4120.mdx:90]: With this sentence I understood: you can use `server:defer` and `client:load` at the same time

```astro
<Foo server:defer client:load />
```
Is that what you wanted to say?

File: src/content/blog/astro-4120.mdx [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+| Source                                                                                                                                         | [WebPageTest](https://www.webpagetest.org/result/240715_BiDc15_G7W/) | [WebPageTest](https://www.webpagetest.org/result/240715_BiDcHW_G7R/) |
+
+As you can see from this table, the numbers are extremely close. Server Islands has an advantage in painting slightly faster due to being more lightweight both in the infrastructure setting in front of the request, and the amount of code needed in the client throughout the process. Partial prerendering has an advantage in that it can finish slightly faster, due to starting the request for its deferred content on the server before the HTML reaches the user's browser.
+
+Overall we are happy with the trade-offs we've made for Server Islands. Our implementation happens mostly at build-time, where component content is swapped out for a small script, one that can almost fit into a Tweet. This means that Server Islands aren't dependant on any particular server infrastructure and will run on host.
```
Comment (delucis) [inline src/content/blog/astro-4120.mdx:106]: Missing word here:

```suggestion
Overall we are happy with the trade-offs we've made for Server Islands. Our implementation happens mostly at build-time, where component content is swapped out for a small script, one that can almost fit into a Tweet. This means that Server Islands aren't dependant on any particular server infrastructure and will run on any host.
```

File: src/content/blog/astro-4120.mdx (hunk 4 above)
Comment (matthewp) [reply src/content/blog/astro-4120.mdx:90]: No, just that you can use it in general. I think there's currently a bug with using them at the same time.

**Comment by FredKSchott** [comment]
reminder to update the pill on the website hero in this PR to Astro 4.12

File: src/content/blog/astro-4120.mdx [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+socialImage: "/src/content/blog/_images/astro-4120/og_image_4_12.webp"
+lang: "en"
+---
+
+import BlogContentImage from "/src/components/BlogContentImage.astro";
+import BlogSummary from "/src/components/BlogSummary.astro";
+import islandDiag from "/src/content/blog/_images/future-of-astro-server-islands/dark-mode-server-islands-diag.webp";
+
+<BlogSummary>
+  Astro 4.12 is out now! This release includes a first look at [Server Islands](https://astro.build/blog/future-of-astro-server-islands/), as well as improvements to pagination and syntax highlighting, and more.
```
Comment (FredKSchott) [inline src/content/blog/astro-4120.mdx:18]: less clear that "first look" means "you can try it today". I'd say "experimental release" instead (or something similar)

Also, add a quick 1-sentence explanation of what Server Islands are. The goal should be that if someone only reads this sentence, they take away the most important bits.

> Astro 4.12 is out now! This release includes our first experimental release of [Server Islands](https://astro.build/blog/future-of-astro-server-islands/) $WITH_TLDR. Improvements to pagination and syntax highlighting are also included.

File: src/content/blog/astro-4120.mdx [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+# Manual:
+npm install astro@latest
+pnpm upgrade astro --latest
+yarn upgrade astro --latest
+```
+
+## Experimental: Server Islands
+
+Today we're announcing experimental support for **Server Islands**, bringing islands architecture to the server. With Server Islands you mark components to run dynamically on the server during page load, allowing the rest of the page to be cached at the edge.
```
Comment (FredKSchott) [inline src/content/blog/astro-4120.mdx:42]: "bringing islands architecture to the server" is a little unclear (I get what you're tring to say here).

Our two messages from last time were "personalization" and "performance", any way to keep hitting those here? Alternatively, a phrase that I liked was the idea that this was "our biggest fundamental change to Islands architecture since the creation of Astro" which is attention grabbing but not super informative. Up to you what to do here!

Comment (FredKSchott) [reply src/content/blog/astro-4120.mdx:42]: > With Server Islands you mark components to run dynamically on the server during page load, allowing the rest of the page to be cached at the edge.

It's not 100% clear how this is different from the way Astro and other web frameworks exist today, to the casual reader. You do a great job at this on the server-islands demo page, can you keep this simple but make it clear why this is a big deal?

File: src/content/blog/astro-4120.mdx [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+
+In 2021 Astro brought the idea of [islands architecture](https://astro.build/blog/introducing-astro/) to the mainstream. We noticed that developers were building content oriented sites using frontend frameworks even though little of the page were dynamic. Islands architecture provides a way to limit the amount of JavaScript loaded in the client without sacrificing the DX (developer experience) that modern frameworks afford.
```
Comment (FredKSchott) [inline src/content/blog/astro-4120.mdx:44]: "We noticed that developers were building content oriented sites using frontend frameworks even though little of the page were dynamic." is hard to parse, honestly paragraph reads better if it's cut entirely or reworded to be the last sentence of the paragraph.

File: src/content/blog/astro-4120.mdx [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+With **Server Islands** we are extending this same architecture to the server side as well. Within any given web page you might have content that is:
```
Comment (FredKSchott) [inline src/content/blog/astro-4120.mdx:46]: "With **Server Islands** we are extending this same architecture to the server side as well." The phrase works well here, but also more reason to not use it in the first paragraph if it exists here

Comment (FredKSchott) [reply src/content/blog/astro-4120.mdx:46]: This part of the post flows funny.

- Sentence 1: Alright, lets get into what Server Islands are
- Sentence 2: Wait, lets go back to talking about your website again

My advice would be to get all of the "state of the web / Astro / your website today" stuff out of the way, and then get into "what are Server Islands". So in that case, cut _"With Server Islands we are extending this same architecture to the server side as well."_ and move it to after the diagram. 

Also, that makes it suddenly more clear that you're spending a lot of time talking about "the way of the world" vs. "server islands" which is what people are really are here to hear an update on. Can you trim down the background content to 1-2 paragraphs + that diagram, and then get to the good stuff about what Server Islands are?

File: src/content/blog/astro-4120.mdx [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+
+<BlogContentImage src={islandDiag} alt="A diagram showing the server island population parts of the page from the server." />
+
+The server islands are used for your most dynamic content; personalized content like the user's avatar, their shopping cart, and product reviews. With these components deferred you're free to more aggressively cache the page itself.
+
+This means that users, logged in or not, will see the most critical parts of the page instantly, as they are caching on Edge CDNs. With careful design considerations the fallback content will be visible for a short amount of time before the dynamic islands are then loaded.
+
+Each island is loaded independent from the rest; this means a slower island, such as one connected to a legacy backend, won't delay the rest of personalized content from being seen and interacted with.
+
+### Trying out Server Islands
```
Comment (FredKSchott) [inline src/content/blog/astro-4120.mdx:62]: I'd put "Performance" section before this. Always best to end with the "how can you try it out today." Plus, Performance section is more aligned with the first section, "how to use it" kind of breaks the flow between them.

File: src/content/blog/astro-4120.mdx [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+
+When Astro builds your site it will omit the component and inject a script in its place, in addition to the content you've marked with `slot="fallback"`. When the page loads in the browser these components will be requested, along with the props you passed to them, to a special endpoint that renders them and returns the HTML.
+
+Server Islands are normal Astro components and can use any of the features you expect from Astro like middleware, slots, `client:` islands.
+
+### Performance
+
+To test the viability of this idea we built [server-islands.com](https://server-islands.com/). This demo is inspired by the [Next.js Partial Prerendering](https://www.partialprerendering.com/) demo, reskinned to suit Astro's personality. **Server Islands** has a lot in common with Partial Prerendering, especially in its goal of allowing increased cacheability of pages.
+
+| <span class="whitespace-nowrap">Metrics</span>                                                                                                 | Partial Prerendering                                                 | Server Islands                                                       |
```
Comment (FredKSchott) [inline src/content/blog/astro-4120.mdx:96]: probably better to use the website domains here, since we're trying to say that this is the performance of the websites and not the technology itself.

File: src/content/blog/astro-4120.mdx [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+| <abbr class="sm:hidden no-underline" title="Total Blocking Time">TBT</abbr><span class="hidden sm:inline">Total Blocking Time</span>           | 0.093s                                                               | 0.018s                                                               |
+| Source                                                                                                                                         | [WebPageTest](https://www.webpagetest.org/result/240715_BiDc15_G7W/) | [WebPageTest](https://www.webpagetest.org/result/240716_AiDc6D_D1N/) |
+
+As you can see from this table, the numbers are extremely close. Server Islands has an advantage in painting slightly faster due to being more lightweight both in the infrastructure setting in front of the request, and the amount of code needed in the client throughout the process. Partial prerendering has an advantage in that it can finish slightly faster, due to starting the request for its deferred content on the server before the HTML reaches the user's browser.
```
Comment (FredKSchott) [inline src/content/blog/astro-4120.mdx:104]: Is the table data right? Feels like SI is faster across the board, 1.251 vs. 1.740 is a 30% speed up. Doesn't match the "extremely close" message at all IMO.

If we're faster, lets celebrate it! You can still couch and caveat but this is the part where we should be bullish and advocate that our approach is superior perf-wise (and why).

File: src/content/blog/astro-4120.mdx [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+
+Overall we are happy with the trade-offs we've made for Server Islands. Our implementation happens mostly at build-time, where component content is swapped out for a small script, one that can almost fit into a Tweet. This means that Server Islands aren't dependant on any particular server infrastructure and will run on any host.
```
Comment (FredKSchott) [inline src/content/blog/astro-4120.mdx:106]: "This means that Server Islands aren't dependant on any particular server infrastructure and will run on any host." 

This is SUPER important, but added here at the end almost as a throw away / an accidental side-effect of our approach. Can the portability angle get its own paragraph?

File: src/content/blog/astro-4120.mdx (hunk 12 above)
Comment (matthewp) [reply src/content/blog/astro-4120.mdx:104]: Good point, reworked this section to be less couch-ey, and moved portability into its own section.

File: src/content/blog/astro-4120.mdx [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+socialImage: "/src/content/blog/_images/astro-4120/og_image_4_12.webp"
+lang: "en"
+---
+
+import BlogContentImage from "/src/components/BlogContentImage.astro";
+import BlogSummary from "/src/components/BlogSummary.astro";
+import islandDiag from "/src/content/blog/_images/future-of-astro-server-islands/dark-mode-server-islands-diag.webp";
+
+<BlogSummary>
+  Astro 4.12 is out now! This release includes a first experimental release of Server Islands, our new solution to integrate high performance static HTML and dynamic server-generated components together. Improvements to pagination and syntax highlighting are also included.
```
Comment (sarah11918) [inline src/content/blog/astro-4120.mdx:18]:
```suggestion
  Astro 4.12 is out now! This release includes the first experimental release of Server Islands, our new solution to integrate high performance static HTML and dynamic server-generated components together. Improvements to pagination and syntax highlighting are also included.
```
Consistent with description above, and sounds a little more natural I think.

File: src/content/blog/astro-4120.mdx [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+
+With **Server islands** we are extending this same architecture to the server. Server islands make it easy to combine high performance static HTML and dynamic server-generated components.
+
+Within any given web page you might have content that is:
+
+- Completely static and never changes.
+- Dynamically backed by a database that changes infrequently, but more often than you deploy.
+- Personalized content, tailored to individual users.
+
+Today you must choose 1 caching strategy for all of these types of content, and if the page is a logged-in experience that usually means no caching at all. With Server islands you get the best of both worlds.
```
Comment (sarah11918) [inline src/content/blog/astro-4120.mdx:52]:
```suggestion
Currently, you must choose one caching strategy for all of these types of content, and if the page is a logged-in experience that usually means no caching at all. Now, with Server islands you get the best of both worlds.
```
(Technically as of *today*, they have a choice! I like using "Now" or "Today" in front of the announcement part to make it sound more like the new thing)

File: src/content/blog/astro-4120.mdx [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+
+<BlogContentImage src={islandDiag} alt="A diagram showing the server island population parts of the page from the server." />
+
+Server islands are used for your most dynamic content; personalized content like the user's avatar, their shopping cart, and product reviews. With these components deferred you're free to more aggressively cache the page itself.
```
Comment (sarah11918) [inline src/content/blog/astro-4120.mdx:56]:
```suggestion
Server islands are used for your most dynamic content; personalized content like a user's avatar, their shopping cart, and product reviews. When these components are deferred, you're free to more aggressively cache the page itself.
```

File: src/content/blog/astro-4120.mdx [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+
+This means that users, logged in or not, will see the most critical parts of the page instantly, as they are caching on Edge CDNs. Fallback content will be visible for a short amount of time before the dynamic islands are then loaded.
+
+Each island is loaded independent from the rest; this means a slower island, such as one connected to a legacy backend, won't delay the rest of personalized content from being seen and interacted with.
+
+### Performance
+
+To test the viability of this idea we built [server-islands.com](https://server-islands.com/). This demo is inspired by the [Next.js Partial Prerendering](https://www.partialprerendering.com/) demo, reskinned to suit Astro's personality. **Server Islands** has a lot in common with Partial Prerendering, especially in its goal of allowing increased cacheability of pages.
```
Comment (sarah11918) [inline src/content/blog/astro-4120.mdx:64]:
```suggestion
To test the viability of this idea we built [server-islands.com](https://server-islands.com/). This demo is inspired by the [Next.js Partial Prerendering](https://www.partialprerendering.com/) demo, reskinned to suit Astro's personality. **Server Islands** has a lot in common with Partial Prerendering (PPR), especially in its goal of allowing increased cacheability of pages.
```
Just mentioning the acronym that you'll be using throughout the rest of the page here, so that it's clear!

File: src/content/blog/astro-4120.mdx [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+
+Server islands, as the name implies, run on the server, so you'll need to use either `'server'` or `'hybrid'` output. Once you've enabled the feature you can use them on any component using the `server:defer` directive:
+
+```astro
+<Avatar server:defer>
+  <GenericUserImage slot="fallback" />
+</Avatar>
+```
+
+When Astro builds your site it will omit the component and inject a script in its place, in addition to the content you've marked with `slot="fallback"`. When the page loads in the browser these components will be requested, along with the props you passed to them, to a special endpoint that renders them and returns the HTML.
```
Comment (sarah11918) [inline src/content/blog/astro-4120.mdx:116]:
```suggestion
When Astro builds your site it will omit the component and inject a script in its place, in addition to the content you've marked with `slot="fallback"`. When the page loads in the browser, these components along with any props you pass to them, will be requested to a special endpoint that renders them and returns the HTML.
```

**Review by sarah11918 — COMMENTED** [review]
Just the tiniest of grammar edits!
